internal/dataplane/fallback/fallback.go

```diff
@@ -31,65 +32,77 @@ func NewGenerator(cacheGraphProvider CacheGraphProvider, logger logr.Logger) *Ge
 	}
 
 	fallbackCache, err := cache.TakeSnapshot()
 	if err != nil {
-		return store.CacheStores{}, fmt.Errorf("failed to take cache snapshot: %w", err)
+		return store.CacheStores{}, nil, fmt.Errorf("failed to take cache snapshot: %w", err)
 	}
 
 	for _, brokenObject := range brokenObjects {
 		subgraphObjects, err := graph.SubgraphObjects(brokenObject)
+		diag = make([]diagnostics.FallbackDiagnostic, len(subgraphObjects))
 		if err != nil {
-			return store.CacheStores{}, fmt.Errorf("failed to find dependants for %s: %w", brokenObject, err)
+			return store.CacheStores{}, nil, fmt.Errorf("failed to find dependants for %s: %w", brokenObject, err)
 		}
-		for _, obj := range subgraphObjects {
+		for i, obj := range subgraphObjects {
 			if err := fallbackCache.Delete(obj); err != nil {
-				return store.CacheStores{}, fmt.Errorf("failed to delete %s from the cache: %w", GetObjectHash(obj), err)
+				return store.CacheStores{}, nil, fmt.Errorf("failed to delete %s from the cache: %w", GetObjectHash(obj), err)
 			}
 			g.logger.V(util.DebugLevel).Info("Excluded object from fallback cache",
 				"object_kind", obj.GetObjectKind(),
 				"object_name", obj.GetName(),
 				"object_namespace", obj.GetNamespace(),
 			)
+			gvk := obj.GetObjectKind().GroupVersionKind()
+			diag[i] = diagnostics.FallbackDiagnostic{
+				GroupKind:     fmt.Sprintf("%s/%s", gvk.Group, gvk.Kind),
+				Namespace:     obj.GetNamespace(),
+				Name:          obj.GetName(),
+				ID:            string(obj.GetUID()),
+				Status:        "excluded",
+				CausingObject: brokenObject.String(),
+			}
 		}
 	}
 
-	return fallbackCache, nil
+	return fallbackCache, diag, nil
 }
 
 func (g *Generator) GenerateBackfillingBrokenObjects(
 	currentCache store.CacheStores,
 	lastValidCacheSnapshot *store.CacheStores,
 	brokenObjects []ObjectHash,
-) (store.CacheStores, error) {
+) (store.CacheStores, []diagnostics.FallbackDiagnostic, error) {
 	// Build a graph from the current cache.
-	currentGraph, err := g.cacheGraphProvider.CacheToGraph(currentCache)
+	currentGraph, diag, err := g.cacheGraphProvider.CacheToGraph(currentCache)
 	if err != nil {
-		return store.CacheStores{}, fmt.Errorf("failed to build current cache graph: %w", err)
+		return store.CacheStores{}, diag, fmt.Errorf("failed to build current cache graph: %w", err)
 	}
 
 	// Take a snapshot of the current cache to use as a fallback.
 	fallbackCache, err := currentCache.TakeSnapshot()
 	if err != nil {
-		return store.CacheStores{}, fmt.Errorf("failed to take current cache snapshot: %w", err)
+		return store.CacheStores{}, diag, fmt.Errorf("failed to take current cache snapshot: %w", err)
 	}
 
 	// Exclude the affected objects from the fallback cache. Also, collect all the affected objects as they will be
 	// subjects of backfilling.
 	var affectedObjects []ObjectHash
+	// Backfill the broken objects from the last valid cache snapshot.
+	var backdiag []diagnostics.FallbackDiagnostic
```

Comment:
Nit: this could be moved to line 126, before the place (`for` loop) where it's used for the first time.